Reset position when out of display bounds

Rexbas opened the pull request:
If the main window is stuck at a display that is no longer attached then the user cannot use the Cryptomator main window, even after restarting. This PR ensures that the position of the window is reset to the primary display after restarting Cryptomator. This happens only if the position stored in `settings.json` is out of display bounds. This PR is related to #3003 and prevents the need for workarounds.

The method `isOutOfDisplayBounds()` checks for each edge of the main window if it is out of bounds. If one of them is out of bounds then the window is reset to the center of the primary display. Or in other words, if a single pixel is out of bounds then the window gets reset.

Rexbas commented:
Maybe it would be good to also check for out of bounds when pressing show in the tray icon.

I also think the method `didDisplayConfigurationChange()` might no longer be necessary as `isOutOfDisplayBounds()` achieves the same but in a different way.

tobihagemann commented on src/main/java/org/cryptomator/ui/mainwindow/ResizeController.java:
```diff
@@ -86,6 +86,39 @@ private boolean didDisplayConfigurationChange() {
 		return configurationHasChanged;
 	}
 
+	private boolean isOutOfDisplayBounds() {
+		final int x = settings.windowXPosition.get();
+		final int y = settings.windowYPosition.get();
+		final int width = settings.windowWidth.get();
+		final int height = settings.windowHeight.get();
+
+		// Left pixel column
+		ObservableList<Screen> screens = Screen.getScreensForRectangle(x, y, 0, height);
+		if (screens.size() < 1) {
+			return true;
+		}
+
+		// Right pixel column
+		screens = Screen.getScreensForRectangle(x + width, y, 0, height);
+		if (screens.size() < 1) {
+			return true;
+		}
+
+		// Top pixel row
+		screens = Screen.getScreensForRectangle(x, y, width, 0);
+		if (screens.size() < 1) {
+			return true;
+		}
+
+		// Bottom pixel row
+		screens = Screen.getScreensForRectangle(x, y + height, width, 0);
+		if (screens.size() < 1) {
+			return true;
+		}
+
+		return false;
+	}
```
Not sure if I've overoptimized this. 😅 And definitely haven't tested it. But maybe that's a more concise solution?

```suggestion
	private boolean isOutOfDisplayBounds() {
		final int x = settings.windowXPosition.get();
		final int y = settings.windowYPosition.get();
		final int width = settings.windowWidth.get();
		final int height = settings.windowHeight.get();

		return isRectangleOutOfScreen(x, y, 0, height) // Left pixel column
				|| isRectangleOutOfScreen(x + width, y, 0, height) // Right pixel column
				|| isRectangleOutOfScreen(x, y, width, 0) // Top pixel row
				|| isRectangleOutOfScreen(x, y + height, width, 0); // Bottom pixel row
	}

	private boolean isRectangleOutOfScreen(int x, int y, int width, int height) {
		return Screen.getScreensForRectangle(x, y, width, height).isEmpty();
	}
```

Rexbas replied:
Yes nice.

Rexbas commented:
I moved the out of bounds check to an event that is triggered when the main window is shown. This ensures the window is always shown within display bounds. I also removed displayConfiguration from settings.

overheadhunter reviewed (changes requested):
Just marking this as "please don't merge yet". Details later, when I am back home :)

overheadhunter commented on src/main/java/org/cryptomator/ui/mainwindow/MainWindowController.java:
```diff
@@ -47,4 +55,37 @@ private void mainWindowFocusChanged(Observable observable) {
+		}
+	}
+
+	private boolean isOutOfDisplayBounds() {
+		final double x = window.getX() + 1;
+		final double y = window.getY() + 1;
+		final double width = window.getWidth() - 2;
+		final double height = window.getHeight() - 2;
+		return isRectangleOutOfScreen(x, y, 0, height) // Left pixel column
+				|| isRectangleOutOfScreen(x + width, y, 0, height) // Right pixel column
+				|| isRectangleOutOfScreen(x, y, width, 0) // Top pixel row
+				|| isRectangleOutOfScreen(x, y + height, width, 0); // Bottom pixel row
+	}
+
+	private boolean isRectangleOutOfScreen(double x, double y, double width, double height) {
+		return Screen.getScreensForRectangle(x, y, width, height).isEmpty();
+	}
```
The intend of `isOutOfDisplayBounds()` seems to be this: Consider a window out of bounds, if at least one of its edges is out of bounds. In other words: The window is out of bounds, if it is not completely within bounds.

This definition is debatable, but for our purpose I'm fine with it. For some fine tuning, however, consider the following example:

![Bounds Check 1](https://github.com/cryptomator/cryptomator/assets/1204330/5d0ab0d1-581b-4d30-a4f7-4952544dcee6)

All green edges are at least partially intersecting at least one screen, even if these edges are just a single pixel in width/height. Shouldn't this be "ouf of bounds", triggering a reset to the center? We could easily "broaden" the edges, needing at least 10px intersections with one of the screens.

Rexbas replied:
The situation in your example should indeed be out of bounds. I think I know another solution. Counting the total visible area of the window and comparing it to the actual area of the window. If the counted area does not match the actual area then it is not within display bounds. I will try it out and test if that works this week. And I agree that `isOutOfDisplayBounds()` might not be the best name. Maybe `isWithinDisplayBounds()` or `isPartiallyOutOfDisplayBounds()` is a better name.

overheadhunter replied:
> The situation in your example should indeed be out of bounds.

As stated above, I'm actually fine with the previous definition. I'm just not sure if this was the intended way. In fact I like the previous implementation far better for its simplicity (and not needing any "double"-based math), however we should make sure to intersect more than just a single pixel column or row. In order to do that, we can inset the window rect by e.g. 10px and feed the four edges of the smaller rect into the "has partial intersection with at least one screen" function.

Remember: The issue we're trying to fix here is that the window is completely out of the screen area, so users can't see it after disconnecting a screen.

Rexbas replied:
So if I am not mistaken, you are saying that the case from your example should not be reset if the width/height of the visible part of the window is more than e.g. 10px?

I think that can easily be done by checking if there are > 0 screens for the blue rectangle. Here the blue rectangle is e.g. 10px smaller in all directions compared to the main window.
![img](https://github.com/cryptomator/cryptomator/assets/14298143/fb6d3178-f884-4a99-b90c-be248ca08d7a)

>The issue we're trying to fix here is that the window is completely out of the screen area, so users can't see it after disconnecting a screen.

This is true but there is also another issue: title bar out of display bounds, making it (near) impossible to move the window around. I thought it would be the easiest to just reset whenever a single pixel is out of bounds but if I understand you correctly you don't agree, which is fine :).

Also, I don't think it is very user friendly if the window is e.g. 90% out of bounds but is not reset because the visible width/height is more than e.g. 10 px. May I suggest using something like 50% of the width for horizontal direction and bottom direction. To make sure the title bar is visible we should do maybe 50% of the height of the title bar in the top of the screen. Percentages because not all screens have the same resolution. This results in something like this: If any of the blue areas are completely out of bounds then reset the window.
![suggested](https://github.com/cryptomator/cryptomator/assets/14298143/47ed087e-b4cb-46d7-9d55-12c070f6ad49)

Before I make any changes please let me know if I understand you correctly and if so, your thoughts about the percentage of the width and height that must be visible to trigger a reset.

overheadhunter replied:
> May I suggest using something like 50% of the width for horizontal direction and bottom direction. To make sure the title bar is visible we should do maybe 50% of the height of the title bar in the top of the screen. Percentages because not all screens have the same resolution.

Yes, this is a reasonable suggestion, since not all areas are equally important when it comes to recovering a partially off-screen window.

We'll also have some offline-discussions about it and I'll chime back in here when we can precisely define the rects.

Rexbas commented:
I pushed new commits and I changed the bounds check method. The method now counts the total visible area of the window. If the visible area is the same as the window area then it is within bounds. I also checked if this works when using duplicate screens mode but in that case `Screen.getScreensForRectangle(...)` returns one screen so the visible area is not counted twice. [This](https://stackoverflow.com/a/52194761) might help understanding the code.

Rexbas commented:
>We'll also have some offline-discussions about it and I'll chime back in here when we can precisely define the rects.

Has this been done?

Anyway, I added commits and made it such that the window only resets when more than 50% of the width or height is out of bounds or when the upper half of the title bar is out of bounds.

infeo commented:
@Rexbas Sorry for keeping this unanswered for so long. I should have picked it up, but missed the notification 🙈 

I might pick up an already answered question, but as [you wrote](https://github.com/cryptomator/cryptomator/issues/3003#issuecomment-1636844743) in the linked issue:
> 4. User disconnects the second display. The window is still located at a position in the second display that is now disconnected.
>5. If the user restarts Cryptomator then didDisplayConfigurationChange will not see any display config changes and will not reset the window to the center of the primary display.

I'm a bit puzzeled, why Cryptomator does not see the change. According to the method `didDisplayConfigurationChange`, we iterate over all screens and store them in the settings with their size. If you detach a screen, you should have either one virtual screen less or its size should have changed. Do I miss something here?

Rexbas commented:
`didDisplayConfigurationChange` only runs once after starting Cryptomator. So if you connect a second display after starting Cryptomator and disconnect it before the next time you start Cryptomator, then it will not detect any changes because there are no changes in display configurations. However, when there was a second display attached temporarily, it was possible to move the Cryptomator window to this second display.

infeo commented on src/main/java/org/cryptomator/ui/mainwindow/ResizeController.java:
```diff
@@ -40,61 +40,56 @@ public class ResizeController implements FxController {
+		// Rect of half the height of the title bar
+		Rectangle2D titleBarRect = new Rectangle2D(window.getX(), window.getY(), window.getWidth(), titleBar.getMinHeight() / 2);
+
+		// Fraction f defines the minimum amount of the width/height that must be visible to not reset the window. Must be less or equal to 0.5
+		double f = 0.5;
+		double minX = window.getX() + f * window.getWidth();
+		double minY = window.getY() + f * window.getHeight();
+		double width = window.getWidth() - 2 * f * window.getWidth();
+		double height = window.getHeight() - 2 * f * window.getHeight();
+		Rectangle2D centerRect = new Rectangle2D(minX, minY, width, height);
```
Don't use fractions here, but a fixed pixel threshold. It should not matter, how big the current window is. Only, if i can see at least XX pixels of of the window.

For the threshold, i suggest to use 10px.

Rexbas replied:
I changed it to 10px but I find it rather small. See the screenshot below. With 10px threshold this Cryptomator window will not be reset but it is barely visible. There is only a small area where you can select the title bar and drag it to a more useable place on the screen. I think this is not very user friendly and the value should be a lot bigger. Could also be different values for width and height.
![image](https://github.com/cryptomator/cryptomator/assets/14298143/87a23b39-a21d-4a08-b5e6-1d88aee23cb3)

infeo commented on src/main/java/org/cryptomator/ui/mainwindow/ResizeController.java:
```diff
@@ -40,61 +40,56 @@ public class ResizeController implements FxController {
+		Rectangle2D centerRect = new Rectangle2D(minX, minY, width, height);
+
+		// Half the title bar or the center rect is out of display bounds
+		if (Screen.getScreensForRectangle(titleBarRect).isEmpty() || Screen.getScreensForRectangle(centerRect).isEmpty()) {
```
With this logic, the following edge case behaves... not entirely as expected?
 Consider a system with the system bar _not_ on the bottom. If i drag the Cryptomator window to the bottom, as soon as i can only see the title bar, it will be reset to primary center on minimize and maximize.
 
 @tobihagemann This okay for you?

Rexbas replied:
Apparently it is not possible to move the system bar anymore in Windows 11 so I can't really reproduce. However, in the version you tested the threshold was set to 50%. So if you move the window to the bottom of the screen, such that only the title bar is visible, then less than 50% is visible so it was reset.

> it will be reset to primary center on minimize and maximize.

You mean minimize to system bar or to tray?
And by maximize you mean fullscreen or restore from minimize?

infeo replied:
This was a logical conclusion, i did not tested it with your branch. The condition says, if either one of the two rectangles (titlebar and main app) is not inside a screen, execute the following bracket scope

> You mean minimize to system bar or to tray?

Should not make a difference, since we added a listener to the `Window.onShowingProperty`. With maximize i meant "show the window again" 😉

infeo commented on src/main/java/org/cryptomator/ui/mainwindow/ResizeController.java:
```diff
@@ -40,61 +40,56 @@ public class ResizeController implements FxController {
-			sb.append("displayId: " + i + ", " + screenBounds.getWidth() + "x" + screenBounds.getHeight() + ";");
+	private void checkDisplayBounds(WindowEvent evt) {
+		// Rect of half the height of the title bar
+		Rectangle2D titleBarRect = new Rectangle2D(window.getX(), window.getY(), window.getWidth(), titleBar.getMinHeight() / 2);
```
Also here i would use a threshold, possibly 15px.

infeo reviewed (changes requested):
Sigh, unfortunately, I misunderstood something when @overheadhunter handed over the issue to me. He explained to me again the direction this PR should go to. I'm sorry to let you jump through some hoops @Rexbas, but we have to rework it 🙈 

So, what _is_ the plan?
Basically, as soon as the app window is not fully contained by a screen, after closing the window (not minimizing!), it should be restored to the default position (primary screen, center). But there is a small catch: We want to allow some slack, i.e. a small portion to be hidden (around 10px, you might know this threshold already 😅).

Implementationwise, it is a merge of https://github.com/cryptomator/cryptomator/pull/3017#discussion_r1264527894, but with aforementioned idea of slack by considering a smaller rectangle (meaning changed x, y positions)

Rexbas commented:
So the window is allowed to be up to 10px out of bounds. When it is more than 10px out of bounds it should reset, right?

infeo commented:
> So the window is allowed to be up to 10px out of bounds. When it is more than 10px out of bounds it should reset, right?

Yes.

Rexbas commented:
I repushed the second version which uses area counting but now with 10px slack, first pushed here: https://github.com/cryptomator/cryptomator/pull/3017#issuecomment-1640744370. Note that the first version, https://github.com/cryptomator/cryptomator/pull/3017#discussion_r1264527894,  does not work well for edge cases like described in https://github.com/cryptomator/cryptomator/pull/3017#discussion_r1265127034.

infeo commented on src/main/java/org/cryptomator/ui/mainwindow/ResizeController.java:
```diff
@@ -53,48 +53,70 @@ public class ResizeController implements FxController {
+		final double height = window.getHeight() - slack;
+		final double x1 = window.getX() + slack;
+		final double y1 = window.getY() + slack;
+		final double x2 = x1 + width;
+		final double y2 = y1 + height;
```
You must substract the slack twice. Otherwise the right bottom corner is the one of the original rectangle, as can be seen mathematically:
```
x2 = x1 + width = (window.getX() + slack) + (window.getWidth() - slack) = window.getX() + window.getWidth()
```

Rexbas replied:
Ah, yes of course.

infeo commented:
Thank you for your contribution and sticking to the PR, @Rexbas!

Rexbas commented:
I think issue #3003 can now be closed, @infeo.

Rexbas commented:
This and #3080 together completely fix #3079, and make sure a minimized window is restored properly (#3080 makes sure it is visible, this PR makes sure it is within bounds if that did not happen with #3080). However this PR's main objective is to fix out of display bounds, which is issue #3003.
